### ISSUE TITLE
Dropbox: error reading destination directory: too_many_requests

### ISSUE THREAD

**Issue by mathieucarbou** [ncw/rclone#1788]
> What is your rclone version (eg output from `rclone -V`)
> Which OS you are using and how many bits (eg Windows 7, 64 bit)

```
rclone v1.38-049-gd0b9baabβ
- os/arch: linux/amd64
- go version: go1.9.2
```

> Which cloud storage system are you using? (eg Google Drive)

`Dropbox`

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`./rclone --log-level "INFO" --transfers 3 --delete-excluded --config "$CONFIG" --exclude-from "EXCLUDES" sync /share/homes/mathieu/Backups dropbox:NAS/Backups`

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
2017/10/31 09:39:58 INFO  : 
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:             10893
Transferred:            0
Elapsed time:        7m1s

2017/10/31 09:40:40 ERROR : mathieu@MACBOOKPRO/mathieu/Pictures/Lightroom/medias/medias Smart Previews.lrdata/8/8AB4: error reading destination directory: too_many_requests/
2017/10/31 09:40:42 ERROR : mathieu@MACBOOKPRO/mathieu/Pictures/Lightroom/medias/medias Smart Previews.lrdata/4/4A14: error reading destination directory: too_many_requests/.
2017/10/31 09:40:46 ERROR : mathieu@MACBOOKPRO/mathieu/Pictures/Lightroom/medias/medias Smart Previews.lrdata/D/D9B3: error reading destination directory: too_many_requests/...
2017/10/31 09:40:48 ERROR : mathieu@MACBOOKPRO/mathieu/Pictures/Lightroom/medias/medias Smart Previews.lrdata/C/C8B8: error reading destination directory: too_many_requests/..
2017/10/31 09:40:58 INFO  : 
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 4
Checks:             10893
Transferred:            0
Elapsed time:        8m1s
```

I think this issue can also be in relation to #1354. Dropbox seems to have a rate-limited API. See:

- https://github.com/saguiitay/DropboxRestAPI/issues/12
- https://www.dropbox.com/developers/documentation/http/documentation

`429 | Your app is making too many requests for the given user or team and is being rate limited. Your app should wait for the number of seconds specified in the "Retry-After" response header before trying again.The Content-Type of the response can be JSON or plaintext. If it is JSON, it will be typeRateLimitErrorYou can find more information in the data ingress guide.
`

It would be nice that rclone obeys this rate limit and retry-after to avoid recording error count, and also log in the console the fact that we are waiting because of the rate limit 😃

**Comment by mathieucarbou** [comment]
Temporary workaround maybe ?

By reading at dropbox documentation here about rate limits (https://www.dropbox.com/developers/reference/data-ingress-guide) it seems that the rate limit is set per oauth token / app linked.

`Rate limits are applied to the OAuth access token used to make the request. For user-linked apps this means rate limits apply per user who has linked your app. Multiple apps that a user may have linked don’t contribute to each other’s rate limiting.`

If we re-issue a `rclone authorize "dropbox"` it regenerate the api token but the app remains the same. 

I was wondering if there could be a trick to have multiple rclone apps inside dropbox to bypass this rate limit and have in the rclone config file several remote sections for dropbox: `dropbox`, `dropbox2`, etc

**Comment by ncw** [comment]
Discussion on another ticket #/1900 indicates that `--tpslimit 9` will fix this problem too.

**Comment by mathieucarbou** [comment]
A PR has been opened (#2622): I have exposed the Retry-After sent by dropbox. But... It is always 300 sec it seems... So it triggers a wait of 5min each time dropbox asks to sleep... Which is quite big... So here are some tests I did with one of my big sync task.

With the latest rclone and `--skip-links --modify-window 2s --log-level INFO --tpslimit 9`: sync time is ~54mins

With the modifications from #2622 and `--skip-links --modify-window 2s --log-level DEBUG`, it uses the Retry-After value that is sent back by dropbox and then we can see some wait during 5 minutes, as seen below in the logs. This same big sync task finishes in ~30mins...

In both cases, no transfer happens, only checks.

So I see value in supporting Retry-After for sync tasks that are short and thus would benefit of a higher request rate, but also that are long, like in this case. So generally speaking, even if Dropbox 5mins wait time seems big, using it instead of a low request rate (`--tpslimit 9`) is better.

```
2018/10/03 15:13:05 DEBUG : ovea/testatoo/.git/objects/67/dd222c58a8ad556c6d3955ef0e98047f12da67: Unchanged skipping
2018/10/03 15:13:05 DEBUG : too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2018/10/03 15:13:05 DEBUG : too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2018/10/03 15:13:05 DEBUG : too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2018/10/03 15:13:05 DEBUG : too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2018/10/03 15:13:38 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:             32905 / 32905, 100%
Transferred:            0 / 0, -
Elapsed time:     16m0.6s

2018/10/03 15:14:38 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:             32905 / 32905, 100%
Transferred:            0 / 0, -
Elapsed time:     17m0.6s

2018/10/03 15:15:38 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:             32905 / 32905, 100%
Transferred:            0 / 0, -
Elapsed time:     18m0.6s

2018/10/03 15:16:38 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:             32905 / 32905, 100%
Transferred:            0 / 0, -
Elapsed time:     19m0.6s

2018/10/03 15:17:38 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:             32905 / 32905, 100%
Transferred:            0 / 0, -
Elapsed time:     20m0.6s

2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 20ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 40ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/..)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 80ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/...)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 160ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 320ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/...)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 640ms
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 1.28s
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/)
2018/10/03 15:18:05 DEBUG : pacer: Rate limited, increasing sleep to 2s
2018/10/03 15:18:05 DEBUG : pacer: low level retry 1/10 (error too_many_requests/...)
2018/10/03 15:18:06 DEBUG : pacer: Reducing sleep to 1.5s
2018/10/03 15:18:06 DEBUG : ovea/pom/.git/objects/e6/9de29bb2d1d6434b8b29ae775ad8c2e48c5391: Size and modification time the same (differ by -370ms, within tolerance 2s)
```

Note: I posted this here in the conversation about his issue because these are good conclusions and tests.

**Comment by davidbitton** [comment]
I'm currently seeing this too. My Dropbox has over 100,000 files and 30TB+ of data. The `sync` was running nicely until I quit it by accident when it was at 95%. Upon re-running the `sync`, rclone chugs through most files quickly because they haven't changed. Then at some point, the rate limit issue comes up, and all retries fail. After 95k files, I start getting:
```
2019-04-25 18:31:09 DEBUG : too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:09 DEBUG : pacer: low level retry 8/10 (error too_many_requests/.)
2019-04-25 18:31:11 DEBUG : too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:11 DEBUG : pacer: low level retry 10/10 (error too_many_requests/..)
2019-04-25 18:31:11 ERROR : PRODUCTS/_requires quick reference codes_/0071 - Flat to Fabulous/WEB: error reading source directory: too_many_requests/..
2019-04-25 18:31:13 DEBUG : too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:13 DEBUG : pacer: low level retry 9/10 (error too_many_requests/..)
2019-04-25 18:31:15 DEBUG : too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:15 DEBUG : pacer: low level retry 10/10 (error too_many_requests/)
2019-04-25 18:31:15 ERROR : PRODUCTS/_requires quick reference codes_/0072 - Light and Airy Backlight/WEB: error reading source directory: too_many_requests/
2019-04-25 18:31:17 DEBUG : too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:17 DEBUG : pacer: low level retry 10/10 (error too_many_requests/.)
2019-04-25 18:31:17 ERROR : PRODUCTS/_requires quick reference codes_/0073 - Fireflies/GRAPHICS: error reading source directory: too_many_requests/.
2019-04-25 18:31:19 DEBUG : too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:19 DEBUG : pacer: low level retry 9/10 (error too_many_requests/...)
2019-04-25 18:31:21 DEBUG : too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:21 DEBUG : pacer: low level retry 8/10 (error too_many_requests/...)
2019-04-25 18:31:23 DEBUG : too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:23 DEBUG : pacer: low level retry 8/10 (error too_many_requests/..)
2019-04-25 18:31:25 DEBUG : too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:25 DEBUG : pacer: low level retry 9/10 (error too_many_requests/...)
2019-04-25 18:31:28 DEBUG : too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:28 DEBUG : pacer: low level retry 1/10 (error too_many_requests/)
2019-04-25 18:31:29 DEBUG : too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2019-04-25 18:31:29 DEBUG : pacer: low level retry 10/10 (error too_many_requests/.)
```

and on and on. Restarting sync right away results in a too_many_request. If I wait some time, I can run it again, but eventually it stops working again. What shall I do?

**Comment by litemotiv** [comment]
I am experiencing this issue and am mainly concerned about an additional problem that it brings, namely that file handles that error out on "too many write operations" are not released which causes the number of open handles to slowly increase. On my debian system this causes rclone to crash with "too many open files" once the number of maximum handles is reached.

`2019/09/17 14:06:43 DEBUG : multisite/wp-content/themes/digikit/vendor/mollie/mollie-api-php/examples/shipments/update-shipment.php: Unchanged skipping                                                 2019/09/17 14:06:43 DEBUG : pacer: low level retry 1/1 (error too_many_write_operations/..)
2019/09/17 14:06:43 DEBUG : pacer: Rate limited, increasing sleep to 20ms
2019/09/17 14:06:43 DEBUG : multisite/wp-content/uploads/sites/66/2014/01/Dorpenlab-Hornhuizen-2.jpg: Received error: upload failed: too_many_write_operations/.. - low level retry 1/10
2019/09/17 14:06:43 DEBUG : pacer: Reducing sleep to 15ms`

Setting the tpslimit (the lowest i've tried is 5) does not seem to prevent the number of handles to slowly rise to the maximum.

**Comment by mathieucarbou** [comment]
@litemotiv @davidbitton  : what are the rclone version and the command-line parameters ?
Also, do you happen do run different rclone process in parallel at the same time or not ?

**Comment by litemotiv** [comment]
@mathieucarbou I'm currently using this version:

`rclone v1.49.0-071-gab89e939-beta
- os/arch: linux/amd64
- go version: go1.13`

Command-line:

`rclone sync -vv -L --log-file=log.txt --tpslimit=5 <source> dropbox:<dest>`

I'm running a single rclone process in a terminal, so far the behaviour has been 100% reproducible on my system.

**Comment by litemotiv** [comment]
@mathieucarbou Since you don't have this issue with open file handles i won't hijack your ticket and will open a new one with a reference to this one.

**Comment by ncw** [comment]
> Since you don't have this issue with open file handles i won't hijack your ticket and will open a new one with a reference to this one.

That would be great - thank you @litemotiv .  If you do a run with `-dump openfiles,goroutines` that will provide useful diagnostics

**Comment by litemotiv** [comment]
> That would be great - thank you @litemotiv . If you do a run with `--dump openfiles,goroutines` that will provide useful diagnostics

Thanks ncw, i assume the dump will only occur when rclone actually crashes right? That can sometimes take a couple of hours depending on the number of errors. Do you know of a way to trigger this earlier, maybe by limiting the number of allowed handles for rclone?

**Comment by nh2** [comment]
Possibly relevant: https://github.com/rclone/rclone/issues/5156#issue-838633716
